Evict failed pods from cache and trigger fresh pod creation.

Bhargav-Singh opened the pull request:
### Problem

In the current implementation, function pods that enter a `Failed` state are not evicted from the cache. Consequently, Porch continues to reuse these failed pods for subsequent evaluations, leading to persistent failures even after the underlying issue (e.g., registry unavailability) has been resolved.

### Solution

This PR introduces the following changes:

* Detects and deletes `Failed` pods during evaluation request handling.
* Deletes and evicts `Failed` pods during garbage collection.
* Adds asynchronous pod deletion (`deleteFailedPod`) with cache safety checks.
* Ensures fresh pod creation is triggered immediately after eviction.
* Adds a unit test case under `TestPodManager` to validate `Failed` pod handling.

efiacor commented:
/ok-to-test

Bhargav-Singh commented:
/retest-required

liamfallon reviewed (commented):
Thanks for authoring this @Bhargav-Singh. It looks good. I wonder if we could get some code coverage on it?

Bhargav-Singh commented:
> Thanks for authoring this @Bhargav-Singh. It looks good. I wonder if we could get some code coverage on it?

Thanks! I’ve added a test (`Failed_pod_is_deleted_and_new_one_is_created`) that exercises the new logic, and it passes. But strangely, the coverage report (both locally and in SonarQube) only shows type definitions as covered, not the actual function code like `podCacheManager()`.

Can I get some help figuring out why is it.

efiacor commented:
> > Thanks for authoring this @Bhargav-Singh. It looks good. I wonder if we could get some code coverage on it?
> 
> Thanks! I’ve added a test (`Failed_pod_is_deleted_and_new_one_is_created`) that exercises the new logic, and it passes. But strangely, the coverage report (both locally and in SonarQube) only shows type definitions as covered, not the actual function code like `podCacheManager()`.
> 
> Can I get some help figuring out why is it.

The unit tests are using fakes so won't provide any coverage metrics. There are some podeval e2e tests. They won't provide coverage either but may be more useful in terms actually of testing the change.

Bhargav-Singh commented:
> > > Thanks for authoring this @Bhargav-Singh. It looks good. I wonder if we could get some code coverage on it?
> > 
> > 
> > Thanks! I’ve added a test (`Failed_pod_is_deleted_and_new_one_is_created`) that exercises the new logic, and it passes. But strangely, the coverage report (both locally and in SonarQube) only shows type definitions as covered, not the actual function code like `podCacheManager()`.
> > Can I get some help figuring out why is it.
> 
> The unit tests are using fakes so won't provide any coverage metrics. There are some podeval e2e tests. They won't provide coverage either but may be more useful in terms actually of testing the change.

I've added e2e test for this. As expected still not getting the code coverage. Is this fine or something more should be done?

efiacor commented on func/internal/podevaluator.go:
```diff
@@ -419,6 +435,19 @@ func (pcm *podCacheManager) garbageCollector() {
 }
 
+// deleteFailedPod deletes a failed pod and removes it from cache if it is still the current pod.
+func (pcm *podCacheManager) deleteFailedPod(pod *corev1.Pod, image string) {
```
Could we move as much as possible to helper functions and test those in isolation?
We have some mocks defined for external interfaces using mockery - https://github.com/nephio-project/porch/tree/main/test/mockery/mocks/external/sigs.k8s.io/controller-runtime/pkg/client

We could add a mock for the grpc client here also.
